linfa-clustering/src/dbscan/algorithm.rs

```diff
@@ -0,0 +1,65 @@
+use crate::dbscan::hyperparameters::DBScanHyperParams;
+use ndarray::{Array1, ArrayBase, Axis, Data, Ix1, Ix2};
+
+pub fn predict(
+    hyperparameters: DBScanHyperParams,
+    observations: &ArrayBase<impl Data<Elem = f64> + Sync, Ix2>,
+) -> Array1<Option<usize>> {
+    let mut result = Array1::from_elem(observations.dim().1, None);
+    let mut latest_id = 0;
+    for (i, obs) in observations.axis_iter(Axis(1)).enumerate() {
+        if result[i].is_some() {
+            continue;
+        }
+        let n = find_neighbors(&obs, observations, hyperparameters.tolerance());
+        if n.len() < hyperparameters.minimum_points() {
+            continue;
+        }
+        // Now go over the neighbours adding them to the cluster
+        let mut search_queue = n
+            .iter()
+            .filter(|x| result[[**x]].is_none())
+            .copied()
+            .collect::<Vec<_>>();
+        while !search_queue.is_empty() {
+            let cand = search_queue.remove(0);
+
+            result[cand] = Some(latest_id);
+
+            let mut n = find_neighbors(&obs, observations, hyperparameters.tolerance())
+                .iter()
+                .filter(|x| result[[**x]].is_none() && !search_queue.contains(x))
+                .copied()
+                .collect::<Vec<_>>();
+
+            search_queue.append(&mut n);
+        }
+        latest_id += 1;
+    }
+    result
+}
```

Comment:
This looks good! 
Improvements we could make from a readability perspective:
- `n` -> `neighbours`
- `cand` -> `candidate`
- `result` -> `cluster_memberships`
- `latest_id` -> `current_cluster_id`

Reply:
Will do, though I've been avoiding writing `neighbours` as it's the UK spelling. Though I'm happy to move everything to the superior UK spelling :wink:

Reply:
I mean, I am not an English native speaker, so I am happy with either as long as its consistent across the crate :stuck_out_tongue: